Fast build check for test

mhutch opened the pull request:
Fixes #5108 Running unit tests aways does full build

mhutch committed:
[Ide] Generalize CheckAndBuildForExecute

This will allow it to be used when executing tests as well

mhutch committed:
[UnitTesting] Use fast build check before running tests

Fixes #5108 Running unit tests aways does full build

Therzok commented on main/src/addins/MonoDevelop.UnitTesting/Services/UnitTestService.cs:
```diff
@@ -126,32 +126,26 @@ internal static Task RunTest (UnitTest test, MonoDevelop.Projects.ExecutionConte
+				await RefreshTests (cs.Token);
 
-					tests = test_names.Select ((fullName) => SearchTest (fullName)).Where ((t) => t != null).ToList ();
+				tests = test_names.Select (SearchTest).Where ((t) => t != null).ToList ();
```
Please revert this change. Roslyn does crappy codegen for method groups.

mhutch replied:
Huh. Is this filed with Roslyn?

Therzok replied:
Yup, long standing too. https://github.com/dotnet/roslyn/issues/5835

Therzok replied:
https://github.com/dotnet/roslyn/issues/3563

mhutch commented on main/src/addins/MonoDevelop.UnitTesting/Services/UnitTestService.cs:
```diff
@@ -126,32 +126,26 @@ internal static Task RunTest (UnitTest test, MonoDevelop.Projects.ExecutionConte
-							return;
-					}
+				var res = await IdeApp.ProjectOperations.CheckAndBuildForExecute (
+					build_targets, ConfigurationSelector.Default, IdeApp.Preferences.BuildBeforeRunningTests,
```
yup

mhutch committed:
[Ide] Improve readability with some code cleanup

Apply fixes for var, pattern matching and event invocation.

mhutch committed:
[Ide] Refactor to expose parallel build functionality

The parallel build functionality from SolutionFolder has been
generalized and moved to Solution.BuildItems and
Solution.CleanItems. Related methods have been moved or obsoleted.

mhutch committed:
[Ide] Parallelize CheckAndBuildForExecute

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/SolutionItem.cs:
```diff
@@ -635,35 +635,16 @@ public Task<BuildResult> Build (ProgressMonitor monitor, ConfigurationSelector s
-
-				SolutionItemConfiguration iconf = GetConfiguration (solutionConfiguration);
-				string confName = iconf != null ? iconf.Id : solutionConfiguration.ToString ();
-				monitor.BeginTask (GettextCatalog.GetString ("Building: {0} ({1})", Name, confName), sortedReferenced.Count);
```
This is changing the current behavior. Solution.BuildItems will log "Building Solution..." and the solution name, instead of "Building..." and the item name.

mhutch replied:
Yup. Probably need Solution.BuildItems to log the name _if it's a single item_ and something generic otherwise?

slluis reviewed (dismissed):
Also, some tests are failing

mhutch committed:
[ProjectTests] Fix tests

ProjectReferencingDisabledProject_ProjectBuildFails was incorrect, when
building in VS the build does not fail.

mhutch committed:
[Ide] Update build/clean task messages

slluis commented on main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/ProjectBuildTests.cs:
```diff
@@ -350,8 +366,7 @@ string GetConfigFolderName (DotNetProject lib, string conf)
 		{
-			string projFile = Util.GetSampleProject ("msbuild-tests", "project-with-custom-build-target2.csproj");
-			var p = (Project)await Services.ProjectService.ReadSolutionItem (Util.GetMonitor (), projFile);
+			(var sol, var p) = await LoadSampleSolutionItem<Project> ("msbuild-tests", "project-with-custom-build-target2.csproj");
```
The solution is not disposed.

slluis commented on main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/ProjectBuildTests.cs:
```diff
@@ -379,8 +394,7 @@ string GetConfigFolderName (DotNetProject lib, string conf)
 		{
-			string projFile = Util.GetSampleProject ("msbuild-tests", "project-with-custom-build-target3.csproj");
-			var p = (Project)await Services.ProjectService.ReadSolutionItem (Util.GetMonitor (), projFile);
+			(var sol, var p) = await LoadSampleSolutionItem<Project> ("msbuild-tests", "project-with-custom-build-target3.csproj");
```
Solution not disposed.

mhutch replied:
It is now :)

slluis commented:
Thanks!
